### PR TITLE
Deny specializing items not in the parent impl

### PR DESCRIPTION
**Pull request by jonas-schievink**
Part of https://github.com/rust-lang/rust/issues/29661 (https://github.com/rust-lang/rfcs/pull/2532). At least sort of?

This was discussed in https://github.com/rust-lang/rust/pull/61812#discussion_r300504114 and is needed for that PR to make progress (fixing an unsoundness).

One annoyance with doing this is that it sometimes requires users to copy-paste a provided trait method into an impl just to mark it `default` (ie. there is no syntax to forward this impl method to the provided trait method).

cc @Centril and @arielb1

### CODE REVIEW

File: src/liballoc/boxed.rs
Comment (Centril) [inline src/liballoc/boxed.rs]: Our current standard library policy is to not expose specialization in public APIs via `default`. Rather, we use an internal trait to mediate the specialization. Can you switch to that?

Also, the implementation could copy https://doc.rust-lang.org/nightly/src/core/iter/traits/iterator.rs.html#269.

File: src/librustc/traits/specialize/specialization_graph.rs
Comment (Centril) [inline src/librustc/traits/specialize/specialization_graph.rs]: Seems like this case `(Type, OpaqueTy)` deserves a comment?

File: src/librustc/traits/specialize/specialization_graph.rs
Comment (Centril) [inline src/librustc/traits/specialize/specialization_graph.rs]: Is this fixme resolved / rendered moot?

File: src/test/ui/specialization/specialization-default-methods.rs
Comment (Centril) [inline src/test/ui/specialization/specialization-default-methods.rs]: It would be good to test the negative case where we don't forward `default fn / type / const` and we get errors.

File: src/librustc/traits/specialize/specialization_graph.rs
Comment (jonas-schievink) [reply src/librustc/traits/specialize/specialization_graph.rs]: Yes, https://github.com/rust-lang/rust/issues/35870 was fixed by #43857

File: src/test/ui/specialization/specialization-default-methods-fail.rs
Comment (Centril) [inline src/test/ui/specialization/specialization-default-methods-fail.rs]: This has no effect :)

File: src/test/ui/specialization/specialization-default-methods-fail.rs
Comment (Centril) [inline src/test/ui/specialization/specialization-default-methods-fail.rs]:
```suggestion
// parent impl causes an error.
```

File: src/test/ui/specialization/specialization-default-methods-fail.rs
Comment (Centril) [inline src/test/ui/specialization/specialization-default-methods-fail.rs]: Looks good -- can this be extended to consts and assoc types?

File: src/test/ui/specialization/specialization-default-methods-fail.rs
Comment (Centril) [inline src/test/ui/specialization/specialization-default-methods-fail.rs]: Also arguably the test should be renamed now. =P (Should have mentioned earlier, sorry! ^^)

**Comment by jonas-schievink** [comment]
Ah I forgot that we wanted to crater this.

@bors try

**Comment by Centril** [comment]
@craterbot run mode=check-only

**Review by nikomatsakis — APPROVED** [review]
r=me once crater run is complete etc

**Comment by jonas-schievink** [comment]
# Summary

There are 7 root regressions: `pyo3`, `tao-of-rust`, `paperclip-core`, `kompact`, `serde_traitobject`, `mockiato` and `incrust`.

There is one seemingly unrelated build failure in [moxie](https://crater-reports.s3.amazonaws.com/pr-64564/try%238d67a66581ea5498d7ac03b8f88a478d8ec75cf0/reg/moxie-0.1.1-alpha.0/log.txt).

# Full Analysis

pyo3
- Axect/freeze_out
- JellyWX/md-html
- RuneBlazer/kmer_vec
- StephanHeijl/global-supertrees
- a-hacker/PyRounders
- aleozlx/sturdy-engine
- ckaran/pyo3_issue_report
- cualbondi/pyosmptparser
- cybermatt/fast-luhn
- dante-signal31/steganer
- deaz/pyo3-test
- gavrie/trial_division
- hhatto/fcsv
- hhatto/fpath
- kdar/rust-python-example
- mozilla/pyo3-parsepatch
- nikhilmitrax/python-oxidation
- omerbenamram/pyevtx-rs
- omerbenamram/pymft-rs
- omerbenamram/pyo3-file
- pattonw/rust-pyn5
- petervaro/rusty-python-fibonacci
- rdeaton/pyskim
- sbdchd/codeowners
- dict_derive
- inline-python-macros
- inline-python
- mbf_gtf
- numpy
- orkhon
- pyo3-built
- pyo3-file
- pyrus-nn
- retworkx
- spellcheck-rs

tao-of-rust

paperclip-core:
- paperclip-actix

kompact:
- Max-Meldrum/kompact_template

serde_traitobject
- alecmocatta/deploy

mockiato
- jeremystucki/mvr
- myelin-ai/engine
- myelin-engine

incrust

[moxie](https://crater-reports.s3.amazonaws.com/pr-64564/try%238d67a66581ea5498d7ac03b8f88a478d8ec75cf0/reg/moxie-0.1.1-alpha.0/log.txt)

**Comment by jonas-schievink** [comment]
Okay, given that there are only 7 root regressions (meaning that only 7 projects need to be fixed after landing this), and this is an unstable feature anyways, I'm going to go ahead and:

@bors r=nikomatsakis

**Pull request merged**
